core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -265,6 +265,10 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     replicaStateMachine.startup()
     partitionStateMachine.startup()
 
+
+    // transition all offline replicas to OfflineReplica state
+    onReplicasBecomeOffline(controllerContext.replicasOnDeadBrokers())
```

Comment:
Hmm.. I wonder if this solves KAFKA-8152 completely. As far as I can tell, the first metadata update sent in `ReplicaStateMachine.startup()` would not have offline partitions included. So there could still be a brief window of inconsistency. It seems we should do both online and offline state transitions before sending any metadata updates?

Reply:
@hachikuji : Yes, I agree that to fix KAFKA-8152, it seems that it's better to fix the logic inside ReplicaStateMachine.startup(). So, in addition to transition all online replicas to ONLINE, we probably want to also transition all offline replicas to OFFLINE to force the propagation of metadata for offline partitions. However, the current logic in ReplicaStateMachine for transitioning to OfflineReplica ignores replicasWithoutLeadershipInfo and therefore won't send LeaderAndIsr and UpdateMetadata requests for those replicas. So, we probably need to fix that part.

Reply:
@junrao @jason Thanks for the review.  I am not sure how to handle [replicasWithoutLeadershipInfo](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/controller/ReplicaStateMachine.scala#L226).  For replicas without leadershipInfo, we don't  have leaderAndIsr in `controllerContext.partitionLeadershipInfo`.  Appreciate your help.

```java
         replicasWithoutLeadershipInfo.foreach { replica =>
          val partition = replica.topicPartition
          if (!topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)) {
            val recipients = controllerContext.partitionReplicaAssignment(partition).filterNot(_ == replicaId)
            val leaderAndIsr = ???
            val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controllerContext.epoch)

            controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients,
              partition,
              leaderIsrAndControllerEpoch,
              controllerContext.partitionReplicaAssignment(partition), isNew = false)
          }
          logSuccessfulTransition(replicaId, replica.topicPartition, replicaState(replica), OfflineReplica)
          replicaState.put(replica, OfflineReplica)
        }
```

Reply:
@junrao @hachikuji Any thoughts on the above?